python-package/lightgbm/basic.py

```diff
@@ -1297,6 +1492,74 @@ def _lazy_init(self, data, label=None, reference=None,
         # set feature names
         return self.set_feature_name(feature_name)
 
+    def __yield_row_from(self, seqs, indices):
+        # type: (List[Sequence], Iterable[int]) -> ...
+        offset = 0
+        seq_id = 0
+        seq = seqs[seq_id]
+        for row_id in indices:
+            assert row_id >= offset, "sample indices are expected to be monotonic"
+            while row_id >= offset + len(seq):
+                offset += len(seq)
+                seq_id += 1
+                seq = seqs[seq_id]
+            id_in_seq = row_id - offset
+            row = seq[int(id_in_seq)]
+            yield row if row.flags['OWNDATA'] else row.copy()
+
+    def __sample(self, seqs, total_nrow):
+        # type: (List[Sequence], int, int) -> (np.ndarray, List[np.ndarray])
+        """Sample data from seqs.
+
+        Mimics behavior in c_api.cpp:LGBM_DatasetCreateFromMats()
+
+        Returns
+        -------
+            sampled_rows, sampled_row_indices
+        """
+        indices = self.create_sample_indices(total_nrow)
+
+        # Select sampled rows, transpose to column order.
+        sampled = [row for row in self.__yield_row_from(seqs, indices)]
+        sampled = np.array(sampled)
+        sampled = sampled.T
+
+        filtered = []
+        filtered_idx = []
+        sampled_row_range = np.arange(len(indices), dtype=np.int32)
+        for col in sampled:
+            col_predicate = (np.abs(col) > ZERO_THRESHOLD) | np.isnan(col)
+            filtered_col = col[col_predicate]
+            filtered_row_idx = sampled_row_range[col_predicate]
+
+            filtered.append(filtered_col)
+            filtered_idx.append(filtered_row_idx)
+
+        return filtered, filtered_idx
+
+    def __init_from_seqs(self, seqs, params_str, ref_dataset):
+        # type: (List[Sequence], str, Dataset) -> None
+        """
+        Initialize data from a Sequence object.
+
+        Sequence: Generic Data Access Object
+            Supports random access and access by batch if properly defined by user
+        """
+        total_nrow = sum(len(seq) for seq in seqs)
+        ncol = len(seqs[0][0])
```

Comment:
Unused variable `ncol`.

Reply:
This was left out for debugging propose and really should be further extended for data sanity check before loading them.

Reply:
However data check could impact performance since `Sequence` interface keeps data access unknown to LightGBM.

I suggest adding a new toggle parameter for checking to control such behavior.

Reply:
Do you mean we need to check that each `seq` has exactly the same number of columns? Doing that here is perfectly ok and won't incur efficiency degradation.

Reply:
> Do you mean we need to check that each `seq` has exactly the same number of columns? 

exactly

> Doing that here is perfectly ok and won't incur efficiency degradation.

Actually we wouldn't know that for sure. 
To exemplify, say our user decided to implement some `seq`s with some very heavy data provider backend and somehow implemented data prefetching and parallelism as their optimization. (In extreme cases imagine data are generated on the fly from some server instances spawned upon our `touching` to the data), 

They might would trigger the generation/read for their whole `seq` when we touch the `seq` since they would expect data are accessed in ordered fashion.

To generalize, we really cannot make assumptions on how much a single access to data would cost. we **hope** they would provide best performance for both **random data access** and **by batch access**.  However in reality, these two are naturally conflicting each other in someway. Assuring batch access is in-order, at least giving an option to do so, is the best we can do to that.

Reply:
> I suggest adding a new parameter for toggling checking behavior.

Actually some helps to this would be very much helpful.

Is it possible to create a `param` for python only, in particular, not to pass down to or cause cpp to warn unrecognized parameter?

Reply:
Yes, that is possible. We need to pick out the parameter in Python before passing it to CPP, which is not an elegant solution. 
Also, we'd better keep the set of parameters in all API consistent.

Reply:
fixed via https://github.com/microsoft/LightGBM/pull/4089/commits/87ef72f1c006a7434f687fd404cd926b86167cbc